src/server/login.js

```diff
@@ -166,37 +166,28 @@ module.exports = function (client, server, options) {
     }
   }
 
-  // https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/util/UUID.java#L163
-  function javaUUID (s) {
-    const hash = crypto.createHash('md5')
-    hash.update(s, 'utf8')
-    const buffer = hash.digest()
-    buffer[6] = (buffer[6] & 0x0f) | 0x30
-    buffer[8] = (buffer[8] & 0x3f) | 0x80
-    return buffer
-  }
-
-  function nameToMcOfflineUUID (name) {
-    return (new UUID(javaUUID('OfflinePlayer:' + name))).toString()
-  }
-
   function loginClient () {
     const isException = !!server.onlineModeExceptions[client.username.toLowerCase()]
     if (onlineMode === false || isException) {
-      client.uuid = nameToMcOfflineUUID(client.username)
+      client.uuid = uuid.nameToMcOfflineUUID(client.username)
     }
     options.beforeLogin?.(client)
     if (client.protocolVersion >= 27) { // 14w28a (27) added whole-protocol compression (http://wiki.vg/Protocol_History#14w28a), earlier versions per-packet compressed TODO: refactor into minecraft-data
       client.write('compress', { threshold: 256 }) // Default threshold is 256
       client.compressionThreshold = 256
     }
+    // TODO: find out what properties are on 'success' packet
     client.write('success', {
       uuid: client.uuid,
       username: client.username,
       properties: []
     })
-    // TODO: find out what properties are on 'success' packet
-    client.state = states.PLAY
+    if (client.supportFeature('hasConfigurationState')) {
+      client.once('login_acknowledged', onClientLoginAck)
+    } else {
+      client.state = states.PLAY
+      server.emit('playerJoin', client)
```

Comment:
this new event should be documented.

It would be breaking for flying-squid (and other users), why do we change from 'login' ?

Reply:
The server sets the mode to from LOGIN to CONFIG after receiving a `login_ack` from the client. The server emits `login` right after the ack, but the state is still in CONFIG so sending PLAY packets (like previous `login` event would let you do) would fail, so in 1.20.2+ now need to either listen for `state` event for it to turn to PLAY or use the added `playerJoin` event before sending game packets

Reply:
Ok makes sense